### PR TITLE
Bluetooth: samples: Add HAP Hearing Aid sample application

### PR DESCRIPTION
**Pull request by MariuszSkamra**
This adds sample application that can be used as reference
implementation and for HAP qualification purposes.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

### CODE REVIEW

File: samples/bluetooth/hap_ha/src/bap_unicast_sr.c [hunk 1]
```diff
@@ -0,0 +1,440 @@
...
+			return source_streams[i].seq_num++;
+		}
+	}
+
+	printk("Could not find endpoint from stream %p\n", stream);
+
+	return 0;
+}
+
+void print_hex(const uint8_t *ptr, size_t len)
```
Comment (Thalley) [inline samples/bluetooth/hap_ha/src/bap_unicast_sr.c:47]:
```suggestion
static void print_hex(const uint8_t *ptr, size_t len)
```

File: samples/bluetooth/hap_ha/src/bap_unicast_sr.c [hunk 2]
```diff
@@ -0,0 +1,440 @@
...
+		       i, codec->data[i].data.type,
+		       codec->data[i].data.data_len);
+		print_hex(codec->data[i].data.data,
+			  codec->data[i].data.data_len -
+			  sizeof(codec->data[i].data.type));
+		printk("\n");
+	}
+
+	if (codec->id == BT_CODEC_LC3_ID) {
+		/* LC3 uses the generic LTV format - other codecs might do as well */
```
Comment (Thalley) [inline samples/bluetooth/hap_ha/src/bap_unicast_sr.c:70]: Isn't this actually required by BAP/CAP?

File: samples/bluetooth/hap_ha/src/ccp_call_ctrl.c [hunk 3]
```diff
@@ -0,0 +1,165 @@
+/** @file
+ *  @brief Bluetooth Telephone Bearer Service client shell
+ *
+ */
```
Comment (Thalley) [inline samples/bluetooth/hap_ha/src/ccp_call_ctrl.c:4]: Not a shell :)

File: samples/bluetooth/hap_ha/src/ccp_call_ctrl.c [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
```
Comment (Thalley) [inline samples/bluetooth/hap_ha/src/ccp_call_ctrl.c:7]: Codecoup?

File: samples/bluetooth/hap_ha/src/csip_set_member.c [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+static void rsi_changed_cb(const uint8_t *rsi)
+{
+	char rsi_str[13];
+
+	snprintk(rsi_str, 163, "%02x%02x%02x%02x%02x%02x",
+		 rsi[0], rsi[1], rsi[2], rsi[3], rsi[4], rsi[5]);
+
+	printk("PRSI: 0x%s\n", rsi_str);
+}
```
Comment (Thalley) [inline samples/bluetooth/hap_ha/src/csip_set_member.c:37]: This should trigger a advertising data update and restart if the sample is advertising the RSI itself, rather than using `bt_csis_advertise`.

File: samples/bluetooth/hap_ha/src/main.c [hunk 6]
```diff
@@ -0,0 +1,191 @@
+/*
+ * Copyright (c) 2021-2022 Nordic Semiconductor ASA
```
Comment (Thalley) [inline samples/bluetooth/hap_ha/src/main.c:2]:
```suggestion
```

File: samples/bluetooth/hap_ha/src/main.c [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+BUILD_ASSERT((CONFIG_BT_PACS_SNK_CONTEXT & MANDATORY_SINK_CONTEXT) == MANDATORY_SINK_CONTEXT,
+	     "Need to support mandatory Supported_Sink_Contexts");
+
+static uint8_t unicast_server_addata[] = {
+	BT_UUID_16_ENCODE(BT_UUID_ASCS_VAL), /* ASCS UUID */
+	BT_AUDIO_UNICAST_ANNOUNCEMENT_TARGETED, /* Target Announcement */
+	(((AVAILABLE_SINK_CONTEXT) >>  0) & 0xFF),
+	(((AVAILABLE_SINK_CONTEXT) >>  8) & 0xFF),
+	(((AVAILABLE_SOURCE_CONTEXT) >>  0) & 0xFF),
+	(((AVAILABLE_SOURCE_CONTEXT) >>  8) & 0xFF),
```
Comment (Thalley) [inline samples/bluetooth/hap_ha/src/main.c:37]: Probably would nice to actually have some `BT_ENCODE_16`, `BT_ENCODE_24`, `BT_ENCODE_32`, etc. macros to handle these (similar to `BT_UUID_16_ENCODE` and `BT_UUID_32_ENCODE`), but that is outside the scope of this PR

File: samples/bluetooth/hap_ha/src/main.c [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+	0x00, /* Metadata length */
+};
+
+/* TODO: Expand with BAP data */
+static const struct bt_data ad[] = {
+	BT_DATA_BYTES(BT_DATA_FLAGS, (BT_LE_AD_GENERAL | BT_LE_AD_NO_BREDR)),
+	BT_DATA_BYTES(BT_DATA_UUID16_ALL, BT_UUID_16_ENCODE(BT_UUID_ASCS_VAL)),
+	BT_DATA(BT_DATA_SVC_DATA16, unicast_server_addata, ARRAY_SIZE(unicast_server_addata)),
+};
```
Comment (Thalley) [inline samples/bluetooth/hap_ha/src/main.c:46]: Looks like it's missing the CSIS data

File: samples/bluetooth/hap_ha/src/main.c [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+
+static struct k_work_delayable adv_work;
+static struct bt_le_ext_adv *adv;
```
Comment (Thalley) [inline samples/bluetooth/hap_ha/src/main.c:49]: I guess this could technically be moved to the `adv_work_handler`

File: samples/bluetooth/hap_ha/src/micp_mic_dev.c [hunk 10]
```diff
@@ -0,0 +1,131 @@
+/** @file
+ *  @brief Bluetooth MICP shell.
```
Comment (Thalley) [inline samples/bluetooth/hap_ha/src/micp_mic_dev.c:2]: Not a shell :)

File: samples/bluetooth/hap_ha/src/micp_mic_dev.c [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+ *
+ * Copyright (c) 2020 Bose Corporation
+ * Copyright (c) 2020-2022 Nordic Semiconductor ASA
```
Comment (Thalley) [inline samples/bluetooth/hap_ha/src/micp_mic_dev.c:5]: Codecoup

File: samples/bluetooth/hap_ha/src/vcp_vol_renderer.c [hunk 12]
```diff
@@ -0,0 +1,180 @@
+/** @file
+ *  @brief Bluetooth VCS server shell.
```
Comment (Thalley) [inline samples/bluetooth/hap_ha/src/vcp_vol_renderer.c:2]: Not a shell :)

File: samples/bluetooth/hap_ha/src/vcp_vol_renderer.c [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+ *
+ * Copyright (c) 2020 Bose Corporation
+ * Copyright (c) 2020-2021 Nordic Semiconductor ASA
```
Comment (Thalley) [inline samples/bluetooth/hap_ha/src/vcp_vol_renderer.c:5]: Codecoup

File: samples/bluetooth/hap_ha/src/vcp_vol_renderer.c [hunk 14]
```diff
@@ -0,0 +1,180 @@
...
+		snprintf(input_desc[i], sizeof(input_desc[i]), "Input %d", i + 1);
+		vcs_param.aics_param[i].description = input_desc[i];
+		vcs_param.aics_param[i].type = BT_AICS_INPUT_TYPE_UNSPECIFIED;
+		vcs_param.aics_param[i].status = true;
+		vcs_param.aics_param[i].gain_mode = BT_AICS_MODE_MANUAL;
+		vcs_param.aics_param[i].units = 1;
+		vcs_param.aics_param[i].min_gain = -100;
+		vcs_param.aics_param[i].max_gain = 100;
+		vcs_param.aics_param[i].cb = &aics_cbs;
+	}
```
Comment (Thalley) [inline samples/bluetooth/hap_ha/src/vcp_vol_renderer.c:162]: Is AICS used for HA?

**Review by Thalley — CHANGES_REQUESTED** [review]
A few minor comments, otherwise LGTM

Very interesting to see a more complete application in Zephyr like this. Hopefully this also gives us some idea of where the current APIs are lacking or are "weird" to use.

**Comment by Thalley** [comment]
@MariuszSkamra For long I've wondered if we could/should add a (or more) "complete" LE Audio setups to the footprint tracking we have in Zephyr (https://stats.zephyrproject.org/testing/d/SSxr5P6Gk/footprint-tracking?orgId=1), so that we can actually start tracking improvements/optimizations over time, and also have a easy way to see the size of such a sample like this. 

Once this is merged, we can consider using this

File: samples/bluetooth/hap_ha/prj.conf [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+CONFIG_BT_EXT_ADV=y
+CONFIG_BT_DEVICE_NAME="Hearing Aid sample"
+CONFIG_BT_DEVICE_APPEARANCE=2625
+
+CONFIG_BT_AUDIO=y
+CONFIG_BT_AUDIO_UNICAST_SERVER=y
+CONFIG_BT_ASCS_ASE_SNK_COUNT=1
+CONFIG_BT_ASCS_ASE_SRC_COUNT=1
+# Support an ISO channel per ASE
+CONFIG_BT_ISO_MAX_CHAN=1
```
[the remark this answers is not in the document]
Comment (MariuszSkamra) [reply samples/bluetooth/hap_ha/prj.conf:18]: Hmm, you have right. The HAP mandates bidirectional, that's why 1 ISO channel. So the comment here is invalid. But anyway, 2 channels should be supported anyway in case the Unicast Client would try to create unidirectional channels. So I'll change it to 2

File: samples/bluetooth/hap_ha/prj.conf [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+# of Volume Offset Control Service (VOCS).
+CONFIG_BT_VOCS_MAX_INSTANCE_COUNT=1
+CONFIG_BT_VCS_VOCS_INSTANCE_COUNT=1
+
+CONFIG_BT_BAS=y
+CONFIG_BT_IAS=y
+CONFIG_BT_CSIS=y
+CONFIG_BT_TBS_CLIENT=y
+
+CONFIG_BT_DEBUG_LOG=y
```
[the remark this answers is not in the document]
Comment (MariuszSkamra) [reply samples/bluetooth/hap_ha/prj.conf:57]: It looks like we do actually

File: samples/bluetooth/hap_ha/src/bap_unicast_sr.c [hunk 17]
```diff
@@ -0,0 +1,440 @@
+/*
+ * Copyright (c) 2021-2022 Nordic Semiconductor ASA
```
[the remark this answers is not in the document]
Comment (MariuszSkamra) [reply samples/bluetooth/hap_ha/src/bap_unicast_sr.c:2]: Nah, I just copied the code from shell app

File: samples/bluetooth/hap_ha/src/vcp_vol_renderer.c (hunk 14 above)
Comment (MariuszSkamra) [reply samples/bluetooth/hap_ha/src/vcp_vol_renderer.c:162]: Yes, here's quote from HAP spec:
"An HA may instantiate one or more instances of Audio Input Control Service (AICS) to expose control of
the gain of its inputs to a Volume Controller."

File: samples/bluetooth/hap_ha/prj.conf (hunk 15 above)
[the remark this answers is not in the document]
Comment (Thalley) [reply samples/bluetooth/hap_ha/prj.conf:18]: One thing to keep in mind:
An ASE is unidirectional, and that 2 ASEs can use the same bidirectional stream (assuming they are a sink and source ASE). 

The relationsship between `CONFIG_BT_ISO_MAX_CHAN` and number of ASEs is pretty loose actually because of this, and the fact that 1 ASE can use up to `CONFIG_BT_MAX_CONN` ISO streams since ASEs are _per connection_ and `CONFIG_BT_ISO_MAX_CHAN` is not. 

IMO it's fine that `CONFIG_BT_ISO_MAX_CHAN=1`, but the comment just stated something different.

File: samples/bluetooth/hap_ha/prj.conf (hunk 16 above)
[the remark this answers is not in the document]
Comment (Thalley) [reply samples/bluetooth/hap_ha/prj.conf:57]: Fair :)

File: samples/bluetooth/hap_ha/src/bap_unicast_sr.c (hunk 17 above)
[the remark this answers is not in the document]
Comment (Thalley) [reply samples/bluetooth/hap_ha/src/bap_unicast_sr.c:2]: Alright. Not sure what the convention here is :)

File: samples/bluetooth/hap_ha/src/csip_set_member.c (hunk 5 above)
Comment (MariuszSkamra) [reply samples/bluetooth/hap_ha/src/csip_set_member.c:37]: Agree, but this requires further refactor in CSIS as well. Now, when `bt_le_ext_adv_cb` has `rpa_expired` callback, it would be good to have a function like `bt_csis_generate_rsi`. That would solve the issue, since, even if application manages the RSI itself, it has to call `bt_csis_advertise` anyway, to get the RSI updates. But if it knows when the RPA expired by `rpa_expired` callback then it would call API function to get new RSI itself

File: samples/bluetooth/hap_ha/prj.conf [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+
+CONFIG_BT_AUDIO=y
+CONFIG_BT_AUDIO_UNICAST_SERVER=y
+CONFIG_BT_ASCS_ASE_SNK_COUNT=1
+CONFIG_BT_ASCS_ASE_SRC_COUNT=1
+# Support an ISO channel per ASE
+CONFIG_BT_ISO_MAX_CHAN=2
+
+CONFIG_BT_PACS_SNK_CONTEXT=0x0047
+CONFIG_BT_PACS_SRC_CONTEXT=0x0003
```
Comment (Thalley) [inline samples/bluetooth/hap_ha/prj.conf:22]: Similarly with the appearance, since this is a sample I guess it would be nice if we had a human-readable format as a comment for these values

File: samples/bluetooth/hap_ha/src/bap_unicast_sr.c [hunk 19]
```diff
@@ -0,0 +1,442 @@
+/** @file
+ *  @brief Bluetooth Call Basic Audio Profile (BAP) Unicast Server role.
```
Comment (Thalley) [inline samples/bluetooth/hap_ha/src/bap_unicast_sr.c:2]:
```suggestion
 *  @brief Bluetooth Basic Audio Profile (BAP) Unicast Server role.
```

File: samples/bluetooth/hap_ha/src/ccp_call_ctrl.c [hunk 20]
```diff
@@ -0,0 +1,163 @@
...
+	}
+
+	return err;
+}
+
+static void connected(struct bt_conn *conn, uint8_t err)
+{
+	if (err) {
+		return;
+	}
```
Comment (Thalley) [inline samples/bluetooth/hap_ha/src/ccp_call_ctrl.c:65]: Should we print something here?

File: samples/bluetooth/hap_ha/src/csip_set_member.c [hunk 21]
```diff
@@ -0,0 +1,49 @@
...
+static struct bt_csis_cb csis_cb = {
+	.lock_changed = csis_lock_changed_cb,
+	.sirk_read_req = sirk_read_req_cb,
+};
+
+int csip_set_member_init(void (*rsi_changed)(const uint8_t *rsi))
+{
+	struct bt_csis_register_param param = {
+		.set_size = 2,
+		.rank = 1,
```
Comment (Thalley) [inline samples/bluetooth/hap_ha/src/csip_set_member.c:40]: If this sample is intended to be used with another device acting as the other hearing aid in the same set, then we will need some way of handling rank. Having 2 devices in the same set with the same rank isn't allowed.

**Review by Thalley — CHANGES_REQUESTED** [review]

File: samples/bluetooth/hap_ha/src/csip_set_member.c (hunk 21 above)
Comment (MariuszSkamra) [reply samples/bluetooth/hap_ha/src/csip_set_member.c:40]: Yes, but my assumption was that someone that would get this code, would assign the rank and SIRK from e.g. flash.
Maybe we could have a test option in kconfig.

Comment (Thalley) [reply samples/bluetooth/hap_ha/src/csip_set_member.c:40]: > Maybe we could have a test option in kconfig.

I think that would be a good idea. That way it's much easier to compile two hearing aids by just supplying something like `-DCONFIG_HA_SET_RANK=1` or `-DCONFIG_HA_SET_RANK=2` to `west build`.

File: samples/bluetooth/hap_ha/prj.conf [hunk 22]
```diff
@@ -0,0 +1,60 @@
...
+
+# If the HA supports the Volume Balance feature (see Section 3.1) and the HA
+# is part of a Binaural Hearing Aid Set, the HA shall instantiate one instance
+# of Volume Offset Control Service (VOCS).
+CONFIG_BT_VOCS_MAX_INSTANCE_COUNT=1
+CONFIG_BT_VCS_VOCS_INSTANCE_COUNT=1
+
+CONFIG_BT_BAS=y
+CONFIG_BT_IAS=y
+CONFIG_BT_CSIS=y
```
Comment (Thalley) [inline samples/bluetooth/hap_ha/prj.conf:54]: Isn't it only for the binaural hearing aid setup that we need to have CSIS? For mono and banded isn't that only a single device, and thus not required to be a set?

Comment (MariuszSkamra) [reply samples/bluetooth/hap_ha/prj.conf:54]: Ah, that it's true. I did not catch that I have it in common .conf file still

**Review by Thalley — DISMISSED** [review]

**Review by Thalley — DISMISSED** [review]

**Comment by MariuszSkamra** [comment]
Updated the codec capabilities to support mandatory 16_2 and 24_2 only.

**Review by Thalley — DISMISSED** [review]

**Comment by MariuszSkamra** [comment]
Improved application-level Kconfig

**Review by jhedberg — APPROVED** [review]

**Review by sjanc — APPROVED** [review]
